allennlp/data/iterators/bucket_iterator.py

```diff
@@ -95,9 +96,117 @@ class BucketIterator(DataIterator):
         If set to `True`, those smaller batches will be discarded.
     """
 
+    def __new__(
+        cls,
+        sorting_keys: List[Tuple[str, str]] = None,
+        padding_noise: float = 0.1,
+        biggest_batch_first: bool = False,
+        batch_size: int = 32,
+        instances_per_epoch: int = None,
+        max_instances_in_memory: int = None,
+        cache_instances: bool = False,
+        track_epoch: bool = False,
+        maximum_samples_per_batch: Tuple[str, int] = None,
+        skip_smaller_batches: bool = False,
+    ):
+
+        dataset_transforms: List[transforms.Transform] = []
+
+        if instances_per_epoch:
+            dataset_transforms.append(transforms.StopAfter(instances_per_epoch))
+
+        if track_epoch:
+            dataset_transforms.append(transforms.EpochTracker())
+
+        if max_instances_in_memory is not None:
+            dataset_transforms.append(transforms.MaxInstancesInMemory(max_instances_in_memory))
+
+        if sorting_keys is not None:
+            # To sort the dataset, it must be indexed.
+            # currently this happens when we call index_with, slightly odd
```

Comment:
Oh - no, it's not different from the old way, it's just a bit more odd in the new functional iterator style. E.g this transform would break, apart from we add in the `Index` transform automatically when we call `iterator.index_with(vocab)`.

Reply:
Maybe reword to something like: "Adding the SortByPadding transform requires that the instances will be indexed by the time the transform is called. Currently this happens when we call index_with in XYZ." I'd drop the part about it being odd unless you want to explain why or suggest an alternative.